DM-50562: Implementation of multi-stage PPDB data ingestion pipeline

JeremyMcCormick opened the pull request:
The main functionality added by this PR is the execution of the "promotion" step for APDB replica chunks, which copies data from staging tables into production. The function for staging chunks was rewritten to copy into the staging tables instead of directly into production.

**Major Changes**

- Added the `promote_chunks` function which promotes APDB replica chunks from staging tables to production
  - This uses the `ReplicaChunkPromoter` in the `db` module of `dax_ppdbx_gcp` for the implementation. 
  - Configuration was included for activating this function on a schedule in the `promote_chunks/schedule.sh` script. The script schedules the promotion process for once a day at noon, Santiago time (Actual execution schedule for production is TBD.).
- Revised the `stage_chunks` function so that it copies data into staging tables rather than directly into production tables.
  - The main updates were to `stage_chunks/stage_chunk_beam_job.py` which implements the Dataflow job for performing this task.
- Added a `track_chunk` function which allows updating the status of a replica chunk using a Pub/Sub event. 
  - This was done mainly to reduce direct dependency on the Postgres database.
  - It is not used uniformly across all functions, as direct DB calls make more sense for certain updates. (The final architecture for this may be modified from what it is, currently.)

**Minor Changes**

- Removed some unnecessary environment variable checks in function scripts and renamed a few variables
- Reworked control flow and error handling in several functions
- General Python code and bash script cleanup, etc.

JeremyMcCormick committed:
Remove unneeded environment variable

JeremyMcCormick committed:
Set Dataflow template path and temp location by convention

Do not read these from the environment as they should not be settable
parameters.

JeremyMcCormick committed:
Update stage_chunk function for compatibility with dax_ppdb changes

The chunk ID is included in the job parameters, and the manifest file
is read from a different location. Some additional log messages were
also added.

JeremyMcCormick committed:
Simplify the Dataflow job parameters and add chunk ID

This also makes a few minor changes to the log messages.

JeremyMcCormick committed:
Add requirements for accessing Postgres

JeremyMcCormick committed:
Add function for tracking replica chunk status

This function inserts or updates a record in the Postgres database used
to track replica chunks.

JeremyMcCormick committed:
Remove unnecessary environment variable checks

Since the script is executing with `set -u` and there is only one
command, the shell will raise errors if variables are undefined when
deploying the function.

JeremyMcCormick committed:
Add teardown script

JeremyMcCormick committed:
Revise track_chunk function to use ppdb-gcp utility modules

JeremyMcCormick committed:
Add transitive dependencies back to work around GCP issues

There are known issues with using transitive dependencies in Cloud Run
functions, so enumerate them all explicitly.

JeremyMcCormick committed:
Write to staging tables and include replica chunk ID

This also includes some revisions to deployment scripts. Some new
utility modules from ppdb-gcp are used, and the chunk tracking db is
updated using a Pub/Sub event.

JeremyMcCormick committed:
Add .env files to .gitignore

JeremyMcCormick committed:
Add initial implementation of cloud function for chunk promotion

This uses utilities in the ppdb-gcp project to promote replica chunks
from staging to production.

JeremyMcCormick committed:
Update ppdb-gcp version to v0.0.8

JeremyMcCormick committed:
Update ppdb-gcp to v0.0.19

JeremyMcCormick committed:
Add daily run scheduling to the promote-chunks function

JeremyMcCormick committed:
Update function requirements to include `dax_ppdbx_gcp`

This also removes some google deps which should be unnecessary. They
should be installed transitively as dependencies of `dax_ppdb_gcp`.

JeremyMcCormick committed:
Update imports with new package name

JeremyMcCormick committed:
Delete the previously scheduled function before rescheduling

JeremyMcCormick committed:
Update lsst-dax-ppdbx-gcp to v0.0.32

JeremyMcCormick committed:
Remove addition of replicaChunkId column as it is in the parquet now

JeremyMcCormick committed:
Add cloud logging config

andy-slac commented on promote_chunks/main.py:
```diff
@@ -0,0 +1,78 @@
+
+    Returns
+    -------
+    Response
```
```suggestion
    response : `Response`
```

andy-slac commented on promote_chunks/main.py:
```diff
@@ -0,0 +1,78 @@
+        promoter.promote_chunks(promotable_chunks)
+
+        # Mark the chunks as promoted in the chunk tracking database
+        promoted_count = _replica_chunk_db.mark_chunks_promoted(promotable_chunks)
```
What happens when this call fails?

JeremyMcCormick replied:
That would be bad if this failed and could lead to data integrity issues if the chunk had the wrong status. In general, making the promotion function idempotent and resilient to errors is challenging. This is one reason to use Pub/Sub over direct db writes because it is easier to manage failures and retries.

If you have any suggestions overall, please let me know. This should be a consideration for future tickets.

andy-slac replied:
> This is one reason to use Pub/Sub over direct db writes because it is easier to manage failures and retries.

I do not believe that, in the end it is still a question about consistency between Postgres and BigQuery tables.

JeremyMcCormick replied:
This is a tough failure mode, because the promotions would not be recorded in the database and the function might attempt to process them again on the next pass. So either the entire promotion should be rolled back, or there should be retries until the database is updated successfully.

Right now, an error should show in the log via:

```
return jsonify({"ok": False, "error": str(e), "chunks_promoted": 0}), 500
```

andy-slac reviewed (approved):
Looks OK to me but I am not a cloud expert, so I was just looking at general things.

JeremyMcCormick committed:
Fix numpydoc type refs

JeremyMcCormick committed:
Update lsst-dax-ppdbx-gcp dep to `main`

JeremyMcCormick committed:
Delete scheduled run when tearing down

JeremyMcCormick committed:
Fix typo in var name

JeremyMcCormick committed:
Add additional pip deps

JeremyMcCormick committed:
Add vpc config to fix db connection errors

JeremyMcCormick committed:
Do not change SA or UA

JeremyMcCormick committed:
Remove `pipefail` from teardown scripts

The `gcloud` commands to delete the functions and other deployment
artifacts may fail if they do not exist, which is fine.
